### ISSUE TITLE
Multiple sub-windows user experience enhancement

### ISSUE THREAD

**Issue by j8xixo12** [solvcon/solvcon#357]
In #356 @yungyuc create a window button in menubar which will use for managing sub-windows.
If multiple  sub-windows are opened in modmesh, they will overlap each other, which is a bad user experience.

The sub-windows management is needed, so create an issue to track it and discussion.

**Comment by tigercosmos** [comment]
@j8xixo12  any solution in mind?

**Comment by yungyuc** [comment]
> @j8xixo12  any solution in mind?

@tigercosmos this is a straight-forward enhancement.  It may be productive to discuss using PR.

**Comment by yungyuc** [comment]
@j8xixo12 Do we need this in the [milestone number 1](https://github.com/solvcon/modmesh/milestone/1), whose goal is to "revamp GUI to provide access to one-dimensional solver and mesh viewing from the menu" (does not include sub-window management)?

**Comment by j8xixo12** [comment]
I think we can put this to next milestone, as the [milestone number 1](https://github.com/solvcon/modmesh/milestone/1) is focusing on one-dimensional solver and mesh viewing. Additionally I expect more user after this release, there will be more GUI issues.
We can solve the following issues together in next milestone.

**Comment by yungyuc** [comment]
Sounds good.  Thanks @j8xixo12

**Comment by yungyuc** [comment]
This seems to be a useful feature.  @j8xixo12 I am moving this to the todo list in https://github.com/orgs/solvcon/projects/8.  We can discuss later whether it should go into the [milestone 1](https://github.com/solvcon/modmesh/milestone/1).

**Comment by yungyuc** [comment]
@YenPeiChen07 Could you please help take a look at this one?

**Comment by YenPeiChen07** [comment]
> @YenPeiChen07 Could you please help take a look at this one?

ok, I will take a look.

**Comment by j8xixo12** [comment]
> @j8xixo12 any solution in mind?

I have a idea, we can use tab to deal with opening multiple applications in the same time, the follwing is my revised version  on figma:
![image](https://github.com/solvcon/modmesh/assets/21517083/f1f9b1e5-6dd7-4fb1-9a9a-abd2d7da327e)

I think the this layout can be used as application template for future application development, so the other developer will not need to take care about GUI code.

**Comment by yungyuc** [comment]
The tab is nice.  I wonder what would be added for the "Windows" menu?  I think the "Tabs" and "Windows" are orthogonal.

**Comment by j8xixo12** [comment]
> The tab is nice. I wonder what would be added for the "Windows" menu? I think the "Tabs" and "Windows" are orthogonal.

I think we can add maximum / minimum main windows size and tile the window to the left / right / top / down side of screen.

**Comment by yungyuc** [comment]
> > The tab is nice. I wonder what would be added for the "Windows" menu? I think the "Tabs" and "Windows" are orthogonal.
> 
> I think we can add maximum / minimum main windows size and tile the window to the left / right / top / down side of screen.

I think we are discussing the management of the sub-windows, not the main window?  We can leave the main window for future work.  Could please list again what should be added in the Windows menu for sub-windows?

**Comment by yungyuc** [comment]
@j8xixo12 @YenPeiChen07 could you please also reference how [paraview](https://www.paraview.org), [gmsh](https://gmsh.info), and [blender](https://www.blender.org) manage sub-windows?

**Comment by j8xixo12** [comment]
> @j8xixo12 @YenPeiChen07 could you please also reference how [paraview](https://www.paraview.org), [gmsh](https://gmsh.info), and [blender](https://www.blender.org) manage sub-windows?

I referred how the paraview, gmsh and blender manage sub-windows. I think we can add 2 types button in Windows menu for sub-windows:
1. **Split plotting window horizontally / vertically**: Allow user to use different view directions to observe the mesh or solution data.
2. **Quick switch sub-windows layout**: Allow user to re-arrange sub-windows in several pre-defined layout.

**Comment by yungyuc** [comment]
Perhaps we could also include a list of sub-windows under the windows menu.  What do you think?

**Comment by j8xixo12** [comment]
@yungyuc I'm not sure why we need a list of sub-windows ?
Is it to make user easier to toggle the sub-windows ?

**Comment by yungyuc** [comment]
Yes, a list of sub-windows helps a user select the current window more easily.

Most multi-window applications or applications supporting sub-windows has a list of windows or sub-windows under the windows menu, e.g., keynotes, safari, mails, blender, etc.

**Comment by j8xixo12** [comment]
> Yes, a list of sub-windows helps a user select the current window more easily.
> 
> Most multi-window applications or applications supporting sub-windows has a list of windows or sub-windows under the windows menu, e.g., keynotes, safari, mails, blender, etc.

Got it, this idea looks good to me!
Let me list the the items we have come up so far here:
1. Split plotting window horizontally / vertically: Allow user to use different view directions to observe the mesh or solution data.
2. Quick switch sub-windows layout: Allow user to re-arrange sub-windows in several pre-defined layout.
3. Add list of sub-windows under `Windows` menu.

@tigercosmos, @YenPeiChen07 Do you have any other idea ?

**Comment by Gene0315** [comment]
For the plotting window, if it is a 3D model, maybe we can set the viewpoint with Azimuth and Elevation. What's more, we can use sliders to adjust the value of angle.
![IMG_8546](https://github.com/user-attachments/assets/a2b9697a-f48a-4634-8107-3fd633e61f14)
![ctrl-sliders-image](https://github.com/user-attachments/assets/7061edb0-2815-4a06-aa1c-d20925c28be7)

**Comment by yungyuc** [comment]
Having a control of the angle will be handy.  We should also reference how other tools (like blender) handle this UI.

**Comment by yungyuc** [comment]
Unassigned for inactivity

**Comment by yungyuc** [comment]
Resolved by PR #1229.
